components/script/dom/bindings/transferable.rs

```diff
@@ -0,0 +1,27 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Trait representing the concept of [transferable objects]
+//! (https://html.spec.whatwg.org/multipage/#transferable-objects).
+use crate::dom::bindings::reflector::DomObject;
+use js::jsapi::{JSContext, JSStructuredCloneReader, MutableHandleObject};
+use std::os::raw;
+
+pub trait Transferable: DomObject {
+    fn transfer(
+        &self,
+        closure: *mut raw::c_void,
+        content: *mut *mut raw::c_void,
+        extra_data: *mut u64,
+    ) -> bool;
```

Comment:
We could simply have it return an `Option<u64>` and do the unsafe part inside the structured-clone callbacks, since that's the only thing currently used?